Improvements of Sum and Product

oscargus opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Improved #16989 
Fixes #17165
Fixes #17256 

#### Brief description of what is fixed or changed
The purpose of this was to speed up the evaluation of Sums and Products. In the meanwhile, a few other issues arised.

The total impact of the execution time may not be enormous. The total test time for test_sums_products is reduced by about a second (from 29), but with some added tests it is back and beyond.

However, on the type of problems I aimed for there is a significant speed-up:
```
from sympy import Sum, sin, cos, Symbol
x = Symbol('x')
h = Symbol('h')
N = Symbol('N', integer=True, positive=True)
%timeit Sum(x*4*cos(h) + 4 + h, (x, 0, N)).doit()
```
Before: `#32.7 ms ± 92.8 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)`
With this PR: `2.6 ms ± 72 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)`

Furthermore, summation variables inherits assumptions from limits when possible. This is a slightly more elaborate/complicated/brute-force approach compared to what is in the Integrals. One may consider if it should be copied there (apart from the integer properties...).

#### Other comments
I also refactored the simplification of Piecewise for later work, although parts of it is used here.

Matrix summations are now directly evaluated in case that the limits are given.

Product first tries some possibly faster methods, even if limits are given.

Apart from regular review comments, I'm also open for suggestions when it comes to naming stuff...

In addition, I fixed a bug when evaluating summations with non-commutative symbols. ~~Should have added a test for it, I realize.~~


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
    * Fixed issue when doing `xreplace` on an expression with `Range`.

* concrete
    * Faster evaluation of certain summations and products.
    * Direct summation of `Sum` with matrices supported.
    * Summation and product variables inherit properties from their limits.
    * Fixed issue where non-commutativity was not honored in `Sum.doit()`.

* simplify
    * The result after `doit()` is only returned if it is smaller than before `doit()`.
    * The content of a `Piecewise` is simplified even when only parts of the expression is a `Piecewise`.

<!-- END RELEASE NOTES -->

oscargus committed:
Added some methods to improve summations and products

oscargus committed:
Fix simplify and tests

oscargus committed:
Fixed bug with using xreplace on ExprWithLimits using Range

oscargus commented on sympy/simplify/simplify.py:
```diff
@@ -585,10 +587,18 @@ def shorter(*choices):
+    if expr.has(Piecewise):
+        expr = piecewise_fold(expr)
+        if expr.has(Piecewise):
+            from sympy.functions.elementary.piecewise import piecewise_simplify
```
This should be removed.

oscargus committed:
Added test for non-commutativity in Sum

oscargus committed:
Added test for xreplace on Sum with Range

oscargus committed:
Fixed issue with inheriting assumptions

oscargus committed:
Fixed issue in piecewise_simplify

oscargus committed:
Updated tests

oscargus commented on sympy/stats/tests/test_discrete_rv.py:
```diff
@@ -217,9 +217,14 @@ def test_conditional():
-        + """(-X2 + n + 3 >= 1) & (-X2 + n + 3 < oo)), (0, True)), (X2, 1, oo), (n, -oo, -1))"""
+    #assert str(P(X1 + X2 < 3, evaluate=False)) == """Sum(Piecewise((2**(X2 - n - 2)*(2/3)**(X2 - 1)/6, """\
+    #    + """(-X2 + n + 3 >= 1) & (-X2 + n + 3 < oo)), (0, True)), (X2, 1, oo), (n, -oo, -1))"""
+    assert str(P(X1 + X2 < 3, evaluate=False)) == 'Sum(Piecewise((2**(-n)/4,'\
```
It seems like the evaluate flag is not propagated here. I cannot really understand why since I didn't touch anything related to that (that I know of). Apart from that, the test should be fine, I think (inheriting assumptions means that the Piecewise relationals are simplified).

smichr replied:
If evaluate is False then is `doit=False` being sent to `simplify` when called somewhere on P's path?

oscargus replied:
It seems a bit weird that simplify should be called upon instantiation. I have not really considered it more than making that statement to myself...

oscarbenjamin replied:
This code should be using Dummys in the sum. We can get rid of the str test by using:
```julia
In [118]: s1 = P(X1 + X2 < 3, evaluate=False)                                                            

In [119]: s2 = Sum(Piecewise((2**(X2 - n - 2)*(S(2)/3)**(X2 - 1)/6, (-X2 + n + 3 >= 1) & (-X2 + n + 3 < o
     ...: o)), (0, True)), (X2, 1, oo), (n, -oo, -1))                                                    

In [120]: s1 = s1.replace(X2.symbol, Dummy('X2', integer=True)).replace(Symbol('n'), Dummy('n', integer=T
     ...: rue))                                                                                          

In [121]: s2 = s2.replace(X2, Dummy('X2', integer=True)).replace(Symbol('n', integer=True), Dummy('n', in
     ...: teger=True))                                                                                   

In [122]: s1.dummy_eq(s2)                                                                                
Out[122]: True
```

oscargus replied:
Agreed. These things showed up at a stage where I had put many more hours than intended on it and since I wasn't sure if I had the correct result, I didn't bother to change it.

oscargus replied:
I can of course change it, once I am reasonably sure that the results are correct (and not just what happened to come out after the code change...).

oscargus replied:
Fixed these now and hope that they are correct.

oscargus commented on sympy/stats/tests/test_discrete_rv.py:
```diff
@@ -217,9 +217,14 @@ def test_conditional():
+        """-X2 + n + 3 >= 1), (0, True)), (X2, 1, oo), (n, 1, oo))"""
+#    assert str(P(Eq(X1 + X2, 3))) == """Sum(Piecewise((2**(X2 - 2)*(2/3)**(X2 - 1)/6, """ +\
+#        """X2 <= 2), (0, True)), (X2, 1, oo))"""
+    assert P(Eq(X1 + X2, 3)) == S(1)/12
```
This one I'm not really sure about. Evaluating the Sum in the original text does not lead to 1/12, but not sure what it should be...

oscargus commented on sympy/simplify/tests/test_simplify.py:
```diff
@@ -805,11 +805,10 @@ def test_issue_15965():
     anew = z*Sum(x**y, (x, 1, a))
     B = Integral(x*y, x)
-    bdo = x**2*y/2
+    bdo = y*Integral(x, x)
```
Apparently y*Integral(x, x) is considered to be a simpler expression than x**2*y/2 (2 vs 3 count_ops).

oscargus committed:
Fixed failing slow tests

smichr commented on sympy/simplify/simplify.py:
```diff
@@ -514,8 +515,18 @@ def simplify(expr, ratio=1.7, measure=count_ops, rational=False, inverse=False,
+
     def done(e):
-        return e.doit() if doit else e
+        return shorter(e.doit(), e) if doit else e
```
As you noticed [here](https://github.com/sympy/sympy/pull/17259/files#r307408318), allowing selection between the doit/undoit form does not respect the doit flag. The docstring says that doit will be called if the flag is not False. If you make it so that the selection is only done if doit is None then a routine can request the shortest and accept the undoit form for compactness sake:
```python
return e.doit() if doit else (shorter(e.doit(), e) if doit is None else e)
```
If so, the docstring should be updated.

oscargus replied:
The reason for this is https://github.com/sympy/sympy/issues/17165#issuecomment-514748083

But you are right that the docstring should be updated. (If it wasn't for the expression growth, I would not have touched it, as I think it makes sense to doit.)

smichr replied:
The concensus was formerly that doit=True was the most user-friendly setting; I would recommend calling with doit=None from Piecewise.

oscargus replied:
OK! Not sure if I follow though. In the internal `_eval_simplify` in Piecewise, I call simplify with doit=None instead of passing the existing doit-flag? (On the expression, I doubt something like that is likely to happen for the conditions.) I guess that will solve the problem, although one cannot force doit on the Piecewise expressions then. But I guess that is the price to pay (and better than the expression growth...). Will try it out.

oscargus replied:
It doesn't work. In the Piecewise specific code, I call simplify on each part and get this:

```
s = (x*Sum(x**n, (n, -1, oo)))
s.doit()
```
![image](https://user-images.githubusercontent.com/8114497/61978054-72876b00-aff0-11e9-9407-7c8f534dbd90.png)
```
s.doit().simplify()
```
![image](https://user-images.githubusercontent.com/8114497/61978069-8206b400-aff0-11e9-9278-9b75b2f4574e.png)
```
s.doit().simplify().simplify()
```
![image](https://user-images.githubusercontent.com/8114497/61978090-93e85700-aff0-11e9-9953-0960882e8dba.png)

From here on it starts looping and keeps constant.

Or did you mean to check if the expression in simplify is a Piecewise and in that case not doit?

oscargus replied:
Oh, and this includes an additional piecewise_fold after the piecewise specific part. So I am quite sure that the final doit is done either way.

oscargus replied:
What about applying a piecewise_fold after the final doit if it contains a Piecewise?

smichr replied:
Maybe...

oscargus replied:
That seems to do the trick. I'm actually wondering if one should have a special path for Piecewise expressions? Not sure if any of the later stages does anything with a Piecewise anyway, so maybe better to just run simplify on the expressions individually (after a piecewise_fold), and then try to simplify it from a Piecewise-perspective (getting rid of pieces).

oscargus committed:
Added final doit back

oscargus committed:
Added final piecewise_fold to avoid Piecewise simplification growth

oscargus committed:
Restored test

oscargus committed:
Use special handling for Piecewise in simplify

oscargus commented:
OK. Now I am trying a dedicated simplify function for Piecewise. As each expression gets a "proper" simplify separately, I am quite sure that this should not have any negative effects, and that the expression growth is stopped.

oscargus committed:
Fixed weird issue in a single ODE test

oscarbenjamin commented:
The weird Dummy comes from here:
https://github.com/sympy/sympy/blob/12c50d4004174407a6afc9ab748ba553de42aa07/sympy/solvers/ode.py#L4109

That part needs fixing for other reasons but the Dummy symbol should have been replaced by constantsimp later...

oscargus commented:
No, not clear why it doesn't replace and why it only happens in a single test. Earlier, it was C + Piecewise. Now, the C has moved into the Piecewise, but one would still expect that the replace should have worked (and it also means that simplify has been called, I would guess, as it is the only reason for piecewise_fold to be called I can think of...)

oscargus committed:
Reverted test

oscargus committed:
Reverted test properly

oscarbenjamin commented:
The weird ODE problem happens here:
https://github.com/sympy/sympy/blob/12c50d4004174407a6afc9ab748ba553de42aa07/sympy/solvers/ode.py#L2970
I think somehow the other changes have caused the constant to be absorbed into the Piecewise so that at this line expr is a Piecewise instead of an Add.

oscargus committed:
Fixed ODE test issue

oscargus commented:
Thanks @oscarbenjamin! That helped me to solve it.

oscargus committed:
Merged with master

smichr commented on sympy/concrete/summations.py:
```diff
@@ -180,11 +180,30 @@ def doit(self, **hints):
+            if d:
+                reps[xab[0]] = d
+        if reps:
+            undo = dict([(v, k) for k, v in reps.items()])
```
I like that variable name.

oscargus replied:
You may probably even recognize it. :-)

smichr replied:
The beauty of forgetting...there are so many little things to see again, anew!

Anyway, I was making a mental not to start using this name after seeing this.

smichr commented on sympy/concrete/summations.py:
```diff
@@ -961,10 +980,62 @@ def eval_sum(f, limits):
 
 def eval_sum_direct(expr, limits):
+    """
+    Evalutate expression directly, but perform some simple checks first
```
```suggestion
    Evaluate expression directly, but perform some simple checks first
```

oscargus committed:
Update sympy/concrete/summations.py

Co-Authored-By: Christopher Smith <smichr@gmail.com>

smichr commented on sympy/core/add.py:
```diff
@@ -352,6 +352,43 @@ def as_coeff_add(self, *deps):
+        If non-commutative, the terms without the deps only include those to
+        the left of the first non-commutative term.
+        """
+        if deps:
```
Would the following work?
```python
c, l2 = self.args_cnc
l1, l2c = sift(c, lambda x: x.has(*deps), binary=True)
return self._from_args(*l1, commutative=True), self._from_args(*(l2c +l2), commutative=False)
```

I am not sure about caching this. Why do you think it is necessary? The other `as_coeff_Foo` routines don't do this.

Instead of creating a new method, can this 3-line version be added as a capability into `as_coeff_Add/Mul` with keyword `deps`?

oscargus replied:
I will have to look into args_cnc, but if it does what I think (in the sense that the code is working), it will, I guess, be working. Just need to make sure that there is a zero and so on.

Caching: `as_coeff_foo` use caching. No idea if it is necessary. Just copied from there.

If the three line version works, it can be used (and even if it doesn't work). Didn't strike me to add a keyword option to `as_coeff_Foo`, but makes sense.

(A two-line version of your code can be used for `as_coeff_foo` as well, I believe.)

oscargus replied:
I used this, but skipped setting the commutative property as I do not think it always should be like that. (It would be to use commutativity for the stuff in l2c (which really should have been the first argument, l1).

oscargus replied:
Much more compact and nicely looking code now, thanks!

smichr commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -1230,3 +1168,87 @@ def _clip(A, B, k):
+    args = []
+    for e, c in expr.args:
+        if isinstance(e, Basic):
+            olddoit = kwargs.pop('doit', None)
```
Since there will be no confusion between a keyword and variable, I would suggest leaving this `doit=` instead of `olddoit`:
```python
>>> def foo(**k): print(k)
>>> doit=1
>>> foo(doit=doit)
{'doit': 1}
```

oscargus replied:
Not sure if I follow. Here, I would like to not use doit (to avoid growth), but for simplification of the conditions I would like to honor the setting. If I just add doit=False and passes on kwargs, there will be a complain about two doit.

Or you just mean the naming of variable? That can be fixed. Not sure what the idea was.

smichr replied:
> just mean the naming of variable

yes. I was demponstrating that there is no harm in using variables that match the assumption keyword. No need for cleverness here...move along. :-)

smichr commented on sympy/solvers/tests/test_ode.py:
```diff
@@ -2287,13 +2287,17 @@ def test_nth_linear_constant_coeff_variation_of_parameters_simplify_False():
     # if this fails because zero.is_zero, replace this block with
     # assert checkodesol(eq, sol_simp, order=5, solve_for_func=False)[0]
-    assert not zero.is_zero and zero.rewrite(exp).simplify() == 0
+    assert not zero.is_zero and zero.rewrite('exp').simplify() == 0
```
Why the use of quotes on `exp`?

oscargus replied:
This code has gone back and forth, so I must have copied the line with quotes when I restored it (and not imported exp when trying it out). Will update.

smichr replied:
I checked...exp is imported in this file.

oscargus replied:
Yes, was probably from when I tried it out by hand in the shell... I've changed it (and the as_coeff-stuff and probably some more). PR soon if the tests pass locally...

smichr commented on sympy/concrete/summations.py:
```diff
@@ -1194,3 +1285,53 @@ def eval_sum_hyper(f, i_a_b):
+                return newf.doit()
+
+
+def dummy_with_inherited_properties(limits):
```
This looks useful. How about
```python
def properties(*values):
    valid = {}
    for a in assumptions or _assume_defined:
        if all(getattr(v, 'is_' + a, None) for v in values):
            valid[a] = True
   return valid
```
As for this routine...like `solveset` the assumptions of the passed symbols should be irrelevant as they are like dummy symbols. So if I do `x = Dummy(negative=True); Integral(x, (x, 1, 3))` I should get the same result as for `Integral(y, (y, 1, 3))`...unless we want to start raising errors if the symbols have assumptions that conflict with the limits. Assuming assumptions are irrelevent (except for this assumption that they are so), you should not keep assumptions which are already True since they may conflict with the assumptions that aren't provided but *are* true based on the limits.

What about:
```python
def consistent_dummy(x, *values):
    valid = properties(*values)
    xa = x._assumptions
    violations = any(k in valid for k for k in x._assumptions if xa[k] is False)
    complete = all(xa[k] for k in valid)
    if violations or not complete:
        return Dummy(x.name, **valid)
```
Neither of these was tested...proceed with caution.

oscargus replied:
> unless we want to start raising errors if the symbols have assumptions that conflict with the limits

I do not see this as a drawback. Except that it actually doesn't happen right now. At least it works well to define variables as integer and still use oo as a limit (which is why I inherited from the provided symbol in the first place, as oo is not integer, there would be no integer assumption on the Dummy, not that I know if it matters, but it makes sense to have integer assumptions on summation variables...)

I'm not sure how your code is supposed to work. How do we provide which assumptions to check for? Provide an additional argument (or two? `_assume_defined`) to properties and consistent_dummy?

I'm sure that there are better ways to write this. I've been busy getting it to work and battling other peculiarities in the current code base. Right now, a new simplify-step that effects Piecewise being merged earlier today...

oscargus committed:
Fixed review comments

oscargus commented:
Thanks @smichr and @oscarbenjamin for review. Updated now with everything(?) except the dummy-assumption stuff. Primarily since I think that the current one solves the problem and I do not really follow how a more generic version should look like (plus that it probably should go in some other file and be used for Integrals, and, well, there has been quite a lot of hassle with caveats showing up here already...)

smichr commented on sympy/concrete/products.py:
```diff
@@ -279,23 +298,31 @@ def _eval_product(self, term, limits):
 
-            for t in term.args:
-                p = self._eval_product(t, (k, a, n))
+            if with_k.is_Mul:
```
Ahh, I was afraid of this and didn't see it before. The use of `as_coeff_[mul|Mul]` depends on what you are going to do with the result. Here you are going to iterate over args so you might as well have used `as_coeff_mul` which has deps built in. OK, but is deps needed on `as_coeff_Mul`? I don't think so. That's what `as_independent` is for, I think. So instead of `foo.as_coeff_Mul(deps=[k])` could you have used `as_independent(k, as_Add=False)`?

smichr replied:
```python
>>> (a*b*X*Y*Z*Y).as_independent(Y)  XYZ noncommutative
(a*b*X, Y*Z*Y)
```

oscargus replied:
The idea was to factor out all the add/mul terms that didn't involve the summation variable and simply take care of those once. The rest of the logic is the same, that if the sum cannot be computed directly, take one term and try to compute the sum for that term and the remaining terms, iterate, so I do not necessarily want to take one term each time (even though I really haven't looked at the code further down, so I do not know if there are any more clever approaches for certain combinations of terms, but I guess that there may be, at least eventually) . I could have used `as_coeff_foo`, but thought it was better to use _from_args than to create new ones (which to some extent is just a matter of definition). I'll look into as_independent and see what it does.

oscargus replied:
That did the trick! Thanks! (It felt like there should have been one, but hard to find.)

oscargus replied:
After a more careful reading of your comment, I realize that you are correct. I'll play around a bit more with as_coeff_mul...

oscargus replied:
My earlier reply was for the summation approach. I had forgotten about this...

oscargus committed:
Used as_independent and renamed dummy generation

oscargus commented on sympy/concrete/summations.py:
```diff
@@ -1194,3 +1285,53 @@ def eval_sum_hyper(f, i_a_b):
+                return newf.doit()
+
+
+def _dummy_with_inherited_properties_concrete(limits):
```
I renamed this. Even with a general function taking assumptions as an argument introduced at a later stage, it would be convenient to still use this as a convenience method for both sum and products to not have to provide the list at every call.

oscargus commented:
Used as_independent and removed all traces of my methods with the same purpose.

I also renamed the dummy generation method to make the purpose more clear. I imagine that it can be used as a convenience function later, if a generic assumption inheritance method is introduced.

oscargus commented:
There is a performance penalty of about 2.5 times using as_independent for the example. Not sure if it is worth adding the dedicated methods back.

oscargus committed:
Used as_coeff_mul instead of as_independent

oscargus committed:
Added some tests for summation with MatrixSymbol

oscargus commented:
OK, there is an issue in `as_independent` that I do not know if I should "solve" or just reintroduce a new function for my purpose. I dread that there will be even more things to fix later if I try to solve it.

```
In [10]: (n*f(x).diff((x, n))).as_independent(n)
Out[10]: (Derivative(f(x), (x, n)), n)
```

It doesn't even come to the actual sorting, but returns here:
https://github.com/sympy/sympy/blob/9406dcdc042780f07debb5bb8575951e9b24789a/sympy/core/expr.py#L1876

oscargus commented:
The problem is really this:
```
In [9]: Derivative(f(x), (x, n)).free_symbols
Out[9]: {x}
```
combined with that `as_independent`redefines `has` internally.

oscargus committed:
Removed empty line at end of file

oscargus committed:
Removed debug printing...

oscargus committed:
Added symbolic derivative count to Derivative.free_symbols

oscargus committed:
Fixed printing

oscargus commented:
I've updated `free_symbols` for `Derivative` to also take into account symbolic derivative numbers. The only test failure was related to printing and require_partial that relied on free_symbols only coming from the expression part. I updated that to do the right thing (only consider the expression), as well as the printing code. So double safety here...

If this passes, it should be good to go from my perspective...

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1131,3 +1157,90 @@ def test_exceptions():
+    assert d.is_real
+    assert d.is_integer
+    assert d.is_nonnegative
+    assert d.is_extended_nonnegative
```
Do you think it should inherit this even though nothing to force extended is present in the limit? i.e. if there is no infinite value should the dummy have the extended enhancement?

oscargus replied:
Good question. I'm not sure it inherits it actually.

```
In [11]: x = Symbol('x', nonnegative=True)

In [12]: x.is_extended_nonnegative
Out[12]: True
```

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -572,23 +572,24 @@ def test_Sum_doit():
-    assert Sum(KroneckerDelta(m, n), (m, -oo, oo)).doit() == Piecewise((1, And(-oo < n, n < oo)), (0, True))
-    assert Sum(x*KroneckerDelta(m, n), (m, -oo, oo)).doit() == Piecewise((x, And(-oo < n, n < oo)), (0, True))
+    # Integer assumes finite
+    # assert Sum(KroneckerDelta(m, n), (m, -oo, oo)).doit() == Piecewise((1, And(-oo < n, n < oo)), (0, True))
```
```suggestion
    assert Sum(KroneckerDelta(x, y), (x, -oo, oo)).doit() == Piecewise((1, And(-oo < y, y < oo)), (0, True))
```

oscargus replied:
```
In [8]: print(Sum(KroneckerDelta(x, y), (x, -oo, oo)).doit())
Piecewise((1, (y >= -oo) & (y < oo)), (0, True))
```

so, no. I'm actually not that keen on adding the "correct" expression either. (Nor fixing the issue so that it returns what may be expected.)

```
In [13]: y = Symbol('y', extended_real=True)

In [14]: print(Sum(KroneckerDelta(x, y), (x, -oo, oo)).doit())
Piecewise((1, y < oo), (0, True))
```

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -572,23 +572,24 @@ def test_Sum_doit():
+    # Integer assumes finite
+    # assert Sum(KroneckerDelta(m, n), (m, -oo, oo)).doit() == Piecewise((1, And(-oo < n, n < oo)), (0, True))
+    assert Sum(KroneckerDelta(m, n), (m, -oo, oo)).doit() == 1
+    # assert Sum(x*KroneckerDelta(m, n), (m, -oo, oo)).doit() == Piecewise((x, And(-oo < n, n < oo)), (0, True))
```
```suggestion
    # assert Sum(m*KroneckerDelta(x, y), (x, -oo, oo)).doit() == Piecewise((m, And(-oo < y, y < oo)), (0, True))
```

oscargus replied:
Similar here:
```
In [12]: print(Sum(m*KroneckerDelta(x, y), (x, -oo, oo)).doit())
Piecewise((m, (y >= -oo) & (y < oo)), (0, True))
```

smichr commented on sympy/core/function.py:
```diff
@@ -1641,7 +1641,11 @@ def derivative_count(self):
 
     @property
     def free_symbols(self):
```
I think this looks right. I was wondering if `Expr.expr_free_symbols` was something related but I don't think so.

oscargus replied:
`Derivative(x, y, evaluate=False).free_symbols` will only return `{x}`, but I think that is probably OK (same as before).

smichr commented on sympy/core/add.py:
```diff
@@ -339,21 +339,18 @@ def as_coeff_add(self, *deps):
         (0, (7*x,))
         """
         if deps:
-            l1 = []
```
I saw someone, perhaps @oscarbenjamin , do this like
```python
L = l1, l2 = [], []
for f in self.args:
    L[f.has(*deps)].append(f)
return self._new_rawargs(*l2), tuple(l1)
```
Since this is a pretty low level function would it make sense to do that and avoid another function call? (Inside `sift` the same thing is being done so all that would be saved is the function call.)

oscargus replied:
Could be. I think the main performance hog is `as_independent` though, and to some extent hiding things into sift is a bit cleaner compared to that code. But then I have no idea how much overhead there is for function calls.

oscargus committed:
Apply suggestions from code review

Co-Authored-By: Christopher Smith <smichr@gmail.com>

oscargus committed:
Merge branch 'master' into sumproductimprovements

oscargus committed:
Merge branch 'sumproductimprovements' of https://github.com/oscargus/sympy into sumproductimprovements

oscargus commented:
I added the DiracDelta tests (against my better knowledge, or maybe gut feeling). Anything else?

oscargus commented:
I'd really try to get this done soon as it is blocking much other work and there are PRs showing up that will probably lead to having to fix tests etc. Is there anything else required apart from sorting out the conflict (and hope that it works after that...).

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1194,3 +1285,53 @@ def eval_sum_hyper(f, i_a_b):
+    assumptions_to_keep = {}
+    assumptions_to_add = {}
+    for assum in assumptions_to_consider:
+        assum_true = x._assumptions.get(assum, None)
```
Is there a reason to use`x._assumptions` rather than `getattr(x, 'is_' +assum)` here? Sometimes assumptions are just defined as class level attributes and I'm not sure that they would end up in `_assumptions`.

oscargus replied:
I think the idea was just to pick the ones that were set and not derive any (as that can wait until they are needed), so this should be more efficient. Also, it was the first way I figured out...

I see your point, but have mainly (only?) tried with expressions where it is set. Note that this will always(?) be a symbol, so for this row I would not expect any issues. Maybe it makes more sense on the next line? Still, it seems like using e.g. integers give a correct answer, without really knowing how the code looks like.

oscarbenjamin commented on sympy/concrete/summations.py:
```diff
@@ -1194,3 +1285,53 @@ def eval_sum_hyper(f, i_a_b):
+def _dummy_with_inherited_properties_concrete(limits):
+    """
+    Return a Dummy symbol that inherits as much assumptions based on the
+    provided symbol and limits as possible.
```
What happens if the provided symbol has assumptions that contradict the limits?

oscargus replied:
Hopefully it throws an inconsistent assumption error? Maybe should be tested.

oscarbenjamin replied:
I tried with this PR:
```julia
In [7]: x = Symbol('x', zero=False)                                                                                               

In [8]: Sum(x, (x, 0, 3)).doit()                                                                                                  
Out[8]: 6

In [9]: x = Symbol('x', positive=True)                                                                                            

In [10]: Sum(x, (x, 0, 3)).doit()                                                                                                 
Out[10]: 6
```

oscargus replied:
Try putting both to zero. It will not add a zero assumption otherwise. I would say that this is consistent with much of SymPy at the moment (also, it would prohibit integer assumptions and summations with `oo`...).

However, I would expect 
```
x = Symbol('x', negative=True) 
Sum(x, (x, 0, 3)).doit() 
```
to throw an error.

oscarbenjamin replied:
I think that the assumptions on the symbol should be ignored. The symbol is a dummy that takes on values determined purely by the limits.

oscargus replied:
The reason that it cannot be ignored (easily) is that it is called recursively and quits if no assumptions are added (similar design as for integrals). I also think it is good practice to put assumptions on variables, rather than relying on automatic assumptions.

Btw, 

```
In [7]: S(1)._assumptions
Out[7]: 
{'commutative': True,
 'integer': True,
 'algebraic': True,
 'noninteger': False,
 'irrational': False,
 'imaginary': False,
 'transcendental': False,
 'extended_real': True,
 'real': True,
 'rational': True,
 'infinite': False,
 'hermitian': True,
 'finite': True,
 'complex': True,
 'zero': False,
 'extended_nonzero': True,
 'nonzero': True}
```

So `negative` or `positive` are not set (except for `oo` etc). Seems to make sense to fix that issue either way.

oscargus replied:
Strange:
```
Integer(1)._assumptions
Out[26]: 
{'commutative': True,
 'integer': True,
 'algebraic': True,
 'noninteger': False,
 'irrational': False,
 'imaginary': False,
 'transcendental': False,
 'extended_real': True,
 'real': True,
 'rational': True,
 'infinite': False,
 'hermitian': True,
 'finite': True,
 'complex': True,
 'zero': False,
 'extended_nonzero': True,
 'nonzero': True,
 'positive': True,
 'negative': False,
 'nonnegative': True,
 'nonpositive': False,
 'extended_nonpositive': False,
 'extended_positive': True,
 'extended_nonnegative': True,
 'extended_negative': False}
````

oscargus replied:
Oh, and a reason to not use the assumptions on the variables is that then one is forced to set assumptions on symbolic bounds. In this way, it is enough to do it on either.

oscarbenjamin replied:
That's why you shouldn't use `_assumptions` directly. It contains the assumptions that have already been checked. When you do `x.is_real` it will check `x._assumptions['real']` then call `x._eval_is_real` and then if that gives True/False update `x._assumptions`. So `_assumptions` is an internal cache.
https://github.com/sympy/sympy/blob/d7a854fb789591114d7d51c3f23bba2fd4d520cc/sympy/core/assumptions.py#L257-L269
If `_eval_is_real` returns None then `_ask` will go on to check any other combination of facts that could imply real e.g. `_eval_is_prime`, `_eval_is_positive`, `_eval_is_extended_real+_eval_is_finite` etc.

oscarbenjamin replied:
> Oh, and a reason to not use the assumptions on the variables is that then one is forced to set assumptions on symbolic bounds. In this way, it is enough to do it on either.

I think that either:
1. It should be checked that the assumptions on both are not known to be inconsistent.
2. The assumptions on the symbol should be ignored.

oscargus commented on sympy/concrete/summations.py:
```diff
@@ -1194,3 +1285,53 @@ def eval_sum_hyper(f, i_a_b):
+        assum_true = x._assumptions.get(assum, None)
+        if assum_true:
+            assumptions_to_keep[assum] = True
+        elif all(i._assumptions.get(assum, None) for i in l):
```
"Next line" = here.

oscargus commented on sympy/concrete/summations.py:
```diff
@@ -1194,3 +1285,53 @@ def eval_sum_hyper(f, i_a_b):
+    If the symbol already has all possible assumptions, return None.
+    Also returns None if no limits are provided.
+    """
+    if len(limits) != 3:
```
This can be removed as all Sums and Products must have explicit limits.

oscargus committed:
Better inconsistent assumptions checking

oscargus committed:
Fixed merge conflicts

oscarbenjamin commented:
This looks good to me. It looks like a lot of good work!

oscarbenjamin commented:
I'm going to merge this so work can proceed. I think that there can still be improvements to the handling of assumptions but there is already #17414 where that can happen.

Good work!
